src/extent.c

```diff
@@ -317,28 +317,31 @@ extents_npages_get(extents_t *extents) {
 }
 
 static void
-extents_insert_locked(tsdn_t *tsdn, extents_t *extents, extent_t *extent) {
+extents_insert_locked(tsdn_t *tsdn, arena_t *arena, extents_t *extents, extent_t *extent) {
 	malloc_mutex_assert_owner(tsdn, &extents->mtx);
-	assert(extent_state_get(extent) == extents->state);
 
 	size_t size = extent_size_get(extent);
 	size_t psz = extent_size_quantize_floor(size);
 	pszind_t pind = sz_psz2ind(psz);
+	mutex_pool_lock(tsdn, &arena->extents_mutex_pool, pind);
```

Comment:
Why not just have NPSIZES mutexes, instead of a whole pool? Pools are pretty big, and doing a 1:1 bin:pool mapping seems strictly better.

Actually, now that I mention it, how do you prevent extent operations from racing with coalescing ones? It seems like you need some per-size-class locks *and* some locks in the existing extent pool.

Reply:
We could, but I didn't want to duplicate the pool code.  We could make the pool size configurable?  

Extent operations racing - The coalesce *does* race, but the can_coalesce function is enough to bail out if something changed.

The activate/deactivate don't really race, because you don't coalesce active extents, and we check for heap empty after locking, so we'd just try the next heap.

The only place I had race issues was activating by address (realloc), where we have to double check after locking nothing has changed.